### PR TITLE
Wzh UI d2 3567 add s3keys config

### PR DESCRIPTION
**Pull request by lizk886**
have a new s3 key model, parser and provider

### CODE REVIEW

File: src/main/java/com/uid2/shared/Const.java [hunk 1]
```diff
@@ -74,6 +74,9 @@ public static class Config {
         public static final String EnforceJwtProp = "enforceJwt";
         public static final String MaaServerBaseUrlProp ="maa_server_base_url";
         public static final String SaltsExpiredShutdownHours = "salts_expired_shutdown_hours";
+        public static final String S3keysMetadataPathProp =  "s3_keys_metadata_path";
+
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/Const.java:78]: Remove extra whitespace please. Also please put these with the other metadata props

File: src/main/java/com/uid2/shared/auth/RotatingOperatorKeyProvider.java [hunk 2]
```diff
@@ -81,8 +81,7 @@ public OperatorKey getOperatorKeyFromHash(String hash) {
 
     @Override
     public Collection<OperatorKey> getAll() {
-        return operatorKeyStore.getAuthorizables();
-    }
+        return operatorKeyStore.getAuthorizables();}
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/auth/RotatingOperatorKeyProvider.java:84]: Please keep proper formatting

File: src/main/java/com/uid2/shared/model/S3Key.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.uid2.shared.model;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/model/S3Key.java:7]: I don't think we need setters for any of these

File: src/main/java/com/uid2/shared/store/parser/S3KeyParser.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class S3KeyParser implements Parser<Map<Integer, S3Key>> {
+    private static final ObjectMapper OBJECT_MAPPER = Mapper.getInstance();
+
+    @Override
+    public ParsingResult<Map<Integer, S3Key>> deserialize(InputStream inputStream) throws IOException {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/parser/S3KeyParser.java:17]: Can you add a test for this

File: src/test/java/com/uid2/shared/store/ScopedStoreReaderTest.java [hunk 5]
```diff
@@ -73,7 +73,6 @@ void returnsLoadedSnapshot() throws Exception {
 
         reader.loadContent(metadata, dataType);
         Collection<TestData> actual = reader.getSnapshot();
-
```
Comment (cody-constine-ttd) [inline src/test/java/com/uid2/shared/store/ScopedStoreReaderTest.java:76]: Unnecessary white space change

File: src/test/java/com/uid2/shared/store/reader/RotatingS3KeyProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,2 @@
+package com.uid2.shared.store.reader;public class RotatingS3KeyProviderTest {
```
Comment (cody-constine-ttd) [inline src/test/java/com/uid2/shared/store/reader/RotatingS3KeyProviderTest.java:1]: Looks like you are still working on this but format it :)

File: src/main/java/com/uid2/shared/model/S3Key.java (hunk 3 above)
Comment (lizk886) [reply src/main/java/com/uid2/shared/model/S3Key.java:7]: don't we need to use setId, set... when we create a new s3 key?

File: src/test/java/com/uid2/shared/store/reader/RotatingS3KeyProviderTest.java (hunk 6 above)
Comment (lizk886) [reply src/test/java/com/uid2/shared/store/reader/RotatingS3KeyProviderTest.java:1]: yes, this is done now!

File: src/main/java/com/uid2/shared/store/parser/S3KeyParser.java (hunk 4 above)
Comment (lizk886) [reply src/main/java/com/uid2/shared/store/parser/S3KeyParser.java:17]: will do

**Review by cody-constine-ttd — APPROVED** [review]
Good once formatting is done

File: src/test/java/com/uid2/shared/store/ScopedStoreReaderTest.java (hunk 5 above)
Comment (lizk886) [reply src/test/java/com/uid2/shared/store/ScopedStoreReaderTest.java]: fixed

File: src/main/java/com/uid2/shared/model/S3Key.java (hunk 3 above)
Comment (lizk886) [reply src/main/java/com/uid2/shared/model/S3Key.java]: Will do, will make all parameters final

File: src/test/java/com/uid2/shared/store/parser/S3KeyParserTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        ParsingResult<Map<Integer, S3Key>> result = parser.deserialize(inputStream);
+
+        assertNotNull(result);
+        assertEquals(3, result.getData().size());
+        assertTrue(result.getData().containsKey(1));
+        assertTrue(result.getData().containsKey(2));
+        assertTrue(result.getData().containsKey(3));
+        assertEquals("S3keySecretByteHere1", result.getData().get(1).getSecret());
+        assertEquals("S3keySecretByteHere2", result.getData().get(2).getSecret());
+        assertEquals("S3keySecretByteHere3", result.getData().get(3).getSecret());
```
Comment (thomasm-ttd) [inline src/test/java/com/uid2/shared/store/parser/S3KeyParserTest.java:43]: You need to assert at least once that all the fields of the s3encryption key has been deserialized correctly. So assert the siteid, activates, created values

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import io.vertx.core.json.JsonObject;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.uid2.shared.model.S3Key;
+
+public class RotatingS3KeyProvider implements StoreReader<Map<Integer, S3Key>> {
+    ScopedStoreReader<Map<Integer, S3Key>> reader;
```
Comment (thomasm-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:18]: this can be final

**Review by cody-constine-ttd — APPROVED** [review]
LGTM

**Pull request merged**